src/mutable_arithmetics.jl

```diff
@@ -0,0 +1,174 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#############################################################################
+# JuMP
+# An algebraic modeling language for Julia
+# See http://github.com/JuliaOpt/JuMP.jl
+#############################################################################
+# src/mutable_arithmetics.jl
+# Implements the mutable arithmetics api defined in MutableArithmetics.jl for
+# `GenericAffExpr` and `GenericQuadExpr`.
+#############################################################################
+
+const _GenericAffOrQuadExpr{C, V} = Union{GenericAffExpr{C, V}, GenericQuadExpr{C, V}}
+
+# The default fallbacks to calling `op(zero(x), zero(y))` which produces allocations.
+# The compiler could avoid these allocations at runtime with constant propagation as the types
+# `x` and `y` are known at compile time but apparently it does not.
+_MA.promote_operation(::Union{typeof(+), typeof(-), typeof(*)}, ::Type{<:_Constant}, V::Type{<:AbstractVariableRef}) = GenericAffExpr{Float64, V}
```

Comment:
These are long lines. Consider using `function` instead of one-liners for better readability.